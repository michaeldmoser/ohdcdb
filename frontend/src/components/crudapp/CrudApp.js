import React, { useState } from 'react';
import { Route, Routes } from 'react-router-dom';

import Context from './context';

import {
    ListOfRecords,
    RecordDetail,
    RecordWithSearchResults,
} from './containers';

const CrudApp = ({ children, useGetListQuery, useGetRecordQuery, ...rest }) => {
    const [search, setSearch] = useState('');
    const [recordId, setRecordId] = useState(null);
    const listQuery = useGetListQuery(search);
    const detailQuery = useGetRecordQuery(recordId, { skip: !recordId });

    const kids = React.Children.toArray(children);
    const recordlist = kids.find(({ type }) => type === ListOfRecords);
    const recordview = kids.find(({ type }) => type === RecordDetail);

    return (
        <Context.Provider
            value={{
                search,
                setSearch,
                recordId,
                setRecordId,
                listQuery,
                detailQuery,
                ...rest,
            }}
        >
            <section className='container-fluid data-view h-100 overflow-hidden'>
                {/* <PropertiesHeader search={search} setSearchQuery={setSearch} /> */}
                <div className='card'>
                    <div className='row h-100 overflow-hidden'>
                        <Routes>
                            {/* <Route
                                index
                                element={<ListOnly>{recordlist}</ListOnly>}
                            /> */}
                            <Route
                                path='*'
                                element={
                                    <RecordWithSearchResults
                                        listview={recordlist}
                                    >
                                        {recordview}
                                    </RecordWithSearchResults>
                                }
                            >
                                {/* <Route path='add' element={<AddProperty />} /> */}
                                <Route path=':recordId' element={recordview} />
                                {/* <Route
                                    path=':recordId/edit'
                                    element={<EditProperty />}
                                /> */}
                            </Route>
                        </Routes>
                    </div>
                </div>
            </section>
        </Context.Provider>
    );
};

export default CrudApp;
