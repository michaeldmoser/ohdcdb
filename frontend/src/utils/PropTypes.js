import PropTypes from 'prop-types';

function isComponent(componentType) {
    const checkComponent = (
        propValue,
        key,
        componentName,
        location,
        propFullName
    ) => {
        if (!propValue[key].type === componentType) {
            return new Error(
                'Invalid prop `' +
                    propFullName +
                    '` supplied to' +
                    ' `' +
                    componentName +
                    '`. Validation failed. Should be a '
            );
        }
    };

    return checkComponent;
}

export default Object.assign({}, PropTypes, { isComponent });
