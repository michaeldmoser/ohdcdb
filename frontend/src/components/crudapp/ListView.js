import { useContext } from 'react';
import PropTypes from 'prop-types';
import { ListGroup } from 'react-bootstrap';
import { NavLink } from 'react-router-dom';

import Context from './context';

const Item = ({ name, additionalInfo }) => {
    return (
        <>
            <h5>{name}</h5>
            <span className='text-muted'>{additionalInfo}</span>
        </>
    );
};

/**
 * Renders a list of records. {children} should be a render function that renders an individual list item
 *
 * @param {array} records A list of the records to render in the list results. It is expectedd that each item in the array is an object with an `id` attribute
 * @param {function} children renderer for the individual list items.
 * @returns JSX
 */
const ListResults = ({ records, children }) => {
    return (
        <ListGroup as='ul' variant='flush'>
            {records.map(({ id, ...rest }) => (
                <ListGroup.Item as='li' className='nav-item' key={id}>
                    <NavLink to={`${id}`} className='lh-sm'>
                        {children({ id, ...rest })}
                    </NavLink>
                </ListGroup.Item>
            ))}
        </ListGroup>
    );
};

ListResults.propTypes = {
    children: PropTypes.func.isRequired,
    records: PropTypes.array.isRequired,
};

const NoResults = () => {
    return <p>No matching records could be found.</p>;
};

const LoadingResults = () => {
    return <p>Loading results...</p>;
};

const Article = ({ listName, children, ...rest }) => {
    return (
        <article className='list-view'>
            <header className='visually-hidden'>
                <h4>{listName || 'List of records'}</h4>
            </header>
            {children}
        </article>
    );
};

/**
 * Basic list view component
 *
 * @param {string} listName The name of the list
 * @param query  The result of an rtk-query useGet.*Query() hook
 * @param children A function which will be passed individual records and should return the rendered record.
 */
export const ListView = ({ listName, children, ...rest }) => {
    const {
        listQuery: { data: records, isLoading },
    } = useContext(Context);

    return (
        <Article listName={listName} {...rest}>
            {isLoading ? (
                <LoadingResults />
            ) : !records ? (
                <NoResults />
            ) : (
                <ListResults records={records}>
                    {(record) => children(record)}
                </ListResults>
            )}
        </Article>
    );
};

ListView.propTypes = {
    listName: PropTypes.string,
    children: PropTypes.func.isRequired,
};
Object.assign(ListView, {
    Item,
});

export default ListView;
