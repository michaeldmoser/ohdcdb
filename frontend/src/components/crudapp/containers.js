import { Children } from 'react';
import { useContext, useEffect } from 'react';
import { useParams, Outlet } from 'react-router-dom';
import PropTypes from 'prop-types';

import Context from './context';

export const RecordDetail = ({ children }) => {
    const { recordId } = useParams();
    const { detailQuery, setRecordId } = useContext(Context);

    useEffect(() => {
        setRecordId(recordId);
    }, [recordId, setRecordId]);

    return children(detailQuery);
};

RecordDetail.propTypes = {
    children: PropTypes.func.isRequired,
};

export const ListOfRecords = ({ children }) => {
    const context = useContext(Context);
    const { recordsList, recordsAreLoading } = context;

    return children({ data: recordsList, isLoading: recordsAreLoading });
};

export const RecordWithSearchResults = ({ listview, children }) => {
    return (
        <>
            <div className='col-3 d-none d-lg-block h-100 overflow-scroll'>
                {listview}
            </div>
            <div className='col-12 col-lg-9'>
                <Outlet />
            </div>
        </>
    );
};
