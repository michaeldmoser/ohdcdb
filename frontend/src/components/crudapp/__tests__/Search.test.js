import React from 'react';
import { render } from 'testing/library';

import { queryResult, database } from './utils';

import CrudApp from '../index';
import { ListOfRecords, RecordDetail } from '../containers';

import ListView, { ListItem } from '../ListView';

describe('Test filtering the list of records', () => {
    const useGetListQuery = (search) => {
        const filtered = search
            ? search.filter((record) => record.title.includes(search))
            : database;

        return queryResult(filtered);
    };

    const useGetRecordQuery = (recordId, options) => {
        return queryResult(options?.skip ? undefined : database[1]);
    };

    xit('should display a filtered list of a records', async () => {
        const details = render(
            <CrudApp {...{ useGetListQuery, useGetRecordQuery }}>
                <ListOfRecords>
                    {(data, isLoading) => {
                        return (
                            <ListView listName='List of records'>
                                {({ title, description }) => (
                                    <ListItem
                                        name={title}
                                        additionalInfo={description}
                                    />
                                )}
                            </ListView>
                        );
                    }}
                </ListOfRecords>
                <RecordDetail>
                    {({ data, isLoading }) => {
                        return <div />;
                    }}
                </RecordDetail>
            </CrudApp>,
            {
                initialEntries: ['/'],
            }
        );

        expect(details).toMatchSnapshot();
    });
});
