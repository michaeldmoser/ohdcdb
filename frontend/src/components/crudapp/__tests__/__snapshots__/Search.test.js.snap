// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test filtering the list of records should display a filtered list of a records 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="container-fluid data-view h-100 overflow-hidden"
      >
        <div
          class="card"
        >
          <div
            class="row h-100 overflow-hidden"
          >
            <div
              class="col-3 d-none d-lg-block h-100 overflow-scroll"
            >
              <article
                class="list-view"
              >
                <header
                  class="visually-hidden"
                >
                  <h4>
                    List of records
                  </h4>
                </header>
                <ul
                  class="list-group list-group-flush"
                >
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/1"
                    >
                      <h5>
                        Schumm Group
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Pre-emptive system-worthy alliance
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/2"
                    >
                      <h5>
                        Wiza - Crist
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Decentralized intangible structure
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/3"
                    >
                      <h5>
                        Stamm LLC
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Multi-channelled 3rd generation attitude
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/4"
                    >
                      <h5>
                        O'Keefe - Funk
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Front-line clear-thinking support
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/5"
                    >
                      <h5>
                        Kling, Koch and Zulauf
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Devolved discrete contingency
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/6"
                    >
                      <h5>
                        Heaney Group
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Vision-oriented reciprocal synergy
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/7"
                    >
                      <h5>
                        Turcotte, Upton and Koch
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Grass-roots tertiary website
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/8"
                    >
                      <h5>
                        Wuckert LLC
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Advanced national core
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/9"
                    >
                      <h5>
                        Emard Group
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Polarised fresh-thinking model
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/10"
                    >
                      <h5>
                        Powlowski, Weber and Mosciski
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Enhanced 24 hour knowledge user
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/11"
                    >
                      <h5>
                        Wolff, Armstrong and Dooley
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Re-contextualized incremental functionalities
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/12"
                    >
                      <h5>
                        Toy and Sons
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Self-enabling intermediate support
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/13"
                    >
                      <h5>
                        Gusikowski - Cummerata
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Business-focused disintermediate superstructure
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/14"
                    >
                      <h5>
                        Bogisich - Greenfelder
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Public-key clear-thinking protocol
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/15"
                    >
                      <h5>
                        Ebert - Mann
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Seamless secondary application
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/16"
                    >
                      <h5>
                        Bartoletti, Baumbach and Smitham
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Switchable solution-oriented monitoring
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/17"
                    >
                      <h5>
                        Ullrich LLC
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Reverse-engineered encompassing monitoring
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/18"
                    >
                      <h5>
                        Hirthe - Fritsch
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Expanded empowering adapter
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/19"
                    >
                      <h5>
                        Roberts, Schmidt and Morar
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Object-based stable paradigm
                      </span>
                    </a>
                  </li>
                  <li
                    class="nav-item list-group-item"
                  >
                    <a
                      class="lh-sm"
                      href="/20"
                    >
                      <h5>
                        Brown, Wolf and Terry
                      </h5>
                      <span
                        class="text-muted"
                      >
                        Customizable methodical intranet
                      </span>
                    </a>
                  </li>
                </ul>
              </article>
            </div>
            <div
              class="col-12 col-lg-9"
            />
          </div>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="container-fluid data-view h-100 overflow-hidden"
    >
      <div
        class="card"
      >
        <div
          class="row h-100 overflow-hidden"
        >
          <div
            class="col-3 d-none d-lg-block h-100 overflow-scroll"
          >
            <article
              class="list-view"
            >
              <header
                class="visually-hidden"
              >
                <h4>
                  List of records
                </h4>
              </header>
              <ul
                class="list-group list-group-flush"
              >
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/1"
                  >
                    <h5>
                      Schumm Group
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Pre-emptive system-worthy alliance
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/2"
                  >
                    <h5>
                      Wiza - Crist
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Decentralized intangible structure
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/3"
                  >
                    <h5>
                      Stamm LLC
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Multi-channelled 3rd generation attitude
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/4"
                  >
                    <h5>
                      O'Keefe - Funk
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Front-line clear-thinking support
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/5"
                  >
                    <h5>
                      Kling, Koch and Zulauf
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Devolved discrete contingency
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/6"
                  >
                    <h5>
                      Heaney Group
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Vision-oriented reciprocal synergy
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/7"
                  >
                    <h5>
                      Turcotte, Upton and Koch
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Grass-roots tertiary website
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/8"
                  >
                    <h5>
                      Wuckert LLC
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Advanced national core
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/9"
                  >
                    <h5>
                      Emard Group
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Polarised fresh-thinking model
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/10"
                  >
                    <h5>
                      Powlowski, Weber and Mosciski
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Enhanced 24 hour knowledge user
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/11"
                  >
                    <h5>
                      Wolff, Armstrong and Dooley
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Re-contextualized incremental functionalities
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/12"
                  >
                    <h5>
                      Toy and Sons
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Self-enabling intermediate support
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/13"
                  >
                    <h5>
                      Gusikowski - Cummerata
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Business-focused disintermediate superstructure
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/14"
                  >
                    <h5>
                      Bogisich - Greenfelder
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Public-key clear-thinking protocol
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/15"
                  >
                    <h5>
                      Ebert - Mann
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Seamless secondary application
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/16"
                  >
                    <h5>
                      Bartoletti, Baumbach and Smitham
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Switchable solution-oriented monitoring
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/17"
                  >
                    <h5>
                      Ullrich LLC
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Reverse-engineered encompassing monitoring
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/18"
                  >
                    <h5>
                      Hirthe - Fritsch
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Expanded empowering adapter
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/19"
                  >
                    <h5>
                      Roberts, Schmidt and Morar
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Object-based stable paradigm
                    </span>
                  </a>
                </li>
                <li
                  class="nav-item list-group-item"
                >
                  <a
                    class="lh-sm"
                    href="/20"
                  >
                    <h5>
                      Brown, Wolf and Terry
                    </h5>
                    <span
                      class="text-muted"
                    >
                      Customizable methodical intranet
                    </span>
                  </a>
                </li>
              </ul>
            </article>
          </div>
          <div
            class="col-12 col-lg-9"
          />
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
